src/Stratis.Features.FederatedPeg/InputConsolidation/InputConsolidator.cs

```diff
@@ -0,0 +1,356 @@
+using NBitcoin;
+using Stratis.Bitcoin.AsyncWork;
+using Stratis.Bitcoin.Features.Wallet.Interfaces;
+using Stratis.Bitcoin.Primitives;
+using Stratis.Bitcoin.Signals;
+using Stratis.Features.FederatedPeg.Events;
+using Stratis.Features.FederatedPeg.Interfaces;
+using Stratis.Features.FederatedPeg.TargetChain;
+using Stratis.Features.FederatedPeg.Wallet;
+
+namespace Stratis.Features.FederatedPeg.InputConsolidation
+{
+    public class InputConsolidator : IInputConsolidator
+    {
+        private readonly IFederationWalletTransactionHandler transactionHandler;
+        private readonly IFederationWalletManager walletManager;
+        private readonly IBroadcasterManager broadcasterManager;
+        private readonly IFederatedPegSettings settings;
+        private readonly ILogger logger;
+        private readonly Network network;
+        private readonly IAsyncProvider asyncProvider;
+
+        /// <summary>
+        /// Queue to handle incoming blocks and update state if needed.
+        /// </summary>
+        private readonly IAsyncDelegateDequeuer<ChainedHeaderBlock> blockQueue;
+
+        /// <summary>
+        /// The task that is building all the consolidation transactions, if one is currently running.
+        /// </summary>
+        private Task consolidationTask;
+
+        /// <summary>
+        /// Used to protect ConsolidationTransactions from write operations.
+        /// </summary>
+        private readonly object lockObj = new object();
+
+        /// <inheritdoc />
+        public List<ConsolidationTransaction> ConsolidationTransactions { get; private set; }
+
+        // TODO: Could put a dictionary by OutPoint.
+
+        public InputConsolidator(IFederationWalletTransactionHandler transactionHandler,
+            IFederationWalletManager walletManager,
+            IBroadcasterManager broadcasterManager,
+            IFederatedPegSettings settings,
+            ILoggerFactory loggerFactory,
+            ISignals signals,
+            IAsyncProvider asyncProvider,
+            Network network)
+        {
+            this.transactionHandler = transactionHandler;
+            this.walletManager = walletManager;
+            this.broadcasterManager = broadcasterManager;
+            this.network = network;
+            this.settings = settings;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+            this.asyncProvider = asyncProvider;
+            this.blockQueue = asyncProvider.CreateAndRunAsyncDelegateDequeuer<ChainedHeaderBlock>($"{nameof(InputConsolidator)}-{nameof(this.blockQueue)}", this.ProcessBlockInternal);
+            signals.Subscribe<WalletNeedsConsolidation>(this.StartConsolidation);
+        }
+
+        /// <inheritdoc />
+        public void StartConsolidation(WalletNeedsConsolidation trigger)
+        {
+            // If there isn't already a task running, start it.
+
+            if (this.consolidationTask == null || this.consolidationTask.IsCompleted)
+            {
```

Comment:
could be a race condition in case StartConsolidation is called 2 times in a row- in that case we will start 2 consolidation tasks

Reply:
Added a lock for task creation